File: mobile_client/app/src/main/java/wpam/mobile_client/protocol/SensorTagSamplesResponse.java
package wpam.mobile_client.protocol;

import java.util.ArrayList;

public final class SensorTagSamplesResponse extends Response
{
    private ArrayList<Integer> values;

    public SensorTagSamplesResponse(ArrayList<Integer> values) {
        super(10, ResponseType.SENSOR_TAG);
        this.values = values;
    }
    public final void close()
    {
        super.close();
    }

    @Override
    public ArrayList<Integer> getFrameBytes()
    {
        initializeDataSize();

        ArrayList<Integer> frame = getHeader();
        frame.add(responseType_.getValue());

        for ( Integer value: values) {
            frame.add(value);
        }

        return frame;
    }

    @Override
    public String getName()
    {
        return new String("BlindsStatusResponse");
    }

    @Override
    public void accept(ResponseVisitor visitor)
    {
        visitor.visit(this);
    }

    public ArrayList<Integer> getValues()
    {
        return this.values;
    }

    public void setValues(ArrayList<Integer> values)
    {
        this.values = values;
    }

    @Override
    protected void initializeDataSize()
    {
        setDataSize(10);
    }
}

File: server/src/protocol/SensorTagSamplesResponse.cpp
#include "SensorTagSamplesResponse.h"

#include "ResponseVisitor.h"
#include <utility/BytesConverter.h>

using namespace std;
using namespace utility;
using namespace communication;

SensorTagSamplesResponse::SensorTagSamplesResponse(const vector<uint16_t > &data)
        : Response(10, ResponseType::SENSOR_TAG),
          data_(data)
{}

SensorTagSamplesResponse::~SensorTagSamplesResponse()
{}

vector<uint8_t> SensorTagSamplesResponse::getFrameBytes()
{
    initializeDataSize();

    vector<uint8_t > frame = getHeader();
    frame.push_back(static_cast<uint8_t>(responseType_));

    for(auto value : data_) {
        BytesConverter::appendUINT16toVectorOfUINT8(value, frame);
    }

    return frame;
}

string SensorTagSamplesResponse::getName()
{
    return string("SensorTagSamplesResponse");
}

void SensorTagSamplesResponse::accept(ResponseVisitor &visitor)
{
    visitor.visit(*this);
}

vector<uint16_t > SensorTagSamplesResponse::getData() const
{
    return data_;
}

void SensorTagSamplesResponse::setMotorStatus(const vector<uint16_t > &data)
{
    data_ = data;
}

void SensorTagSamplesResponse::initializeDataSize()
{
    uint16_t dataSize = sizeof(responseType_);
    dataSize += sizeof(uint16_t)*data_.size();

    setDataSize(dataSize);
}
